### ISSUE TITLE
OpenIdConnect: setting the ClaimsIssuer property in configuration options has no effect

### ISSUE THREAD

**Issue by amoszheng** [dotnet/aspnetcore#41589]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

When using the package Microsoft.AspNetCore.Authentication.OpenIdConnect in an ASP.NET Core application to add external login providers supporting OpenId Connect protocol, setting the `ClaimsIssuer` property in `OpenIdConnectOptions` has no effect; the principal claims still get generated with the issuer that comes from the external identity provider. 

This behavior is in contrast with other social media login providers (Microsoft Account, Facebook, Google, etc.) where specifying this property in the configuration options would cause the principal claims to be issued with the specified claims issuer. 

Is is possible to fix that so the `ClaimsIssuer` option in the OpenId Connect client works the same way it does in other social login provider packages? That way we can use it for any external identity provider that supports the OpenId Connect protocol but does not have a specific package built for it.

### Expected Behavior

Once the `ClaimsIssuer` property of the `OpenIdConnectOptions` is set, the principal claims should be issued with the specified claims issuer.

### Steps To Reproduce

1. Configure the OpenIdConnect client with any identity provider that supports OpenId Connect;
2. Set the `ClaimsIssuer` property to something different (e.g. "MyCustomIssuer") in `.AddOpenIdConnect()` configuration;
3. Verify that the actual issuer of the principal claims is unchanged.

**Comment by adityamandaleeka** [comment]
Let's investigate and see whether this is an IdentityModel limitation or we just need to pass something through.

**Comment by Tratcher** [comment]
ClaimsIssuer is used by implementations that create Claims from sources that don't provide that information. E.g. 
https://github.com/dotnet/aspnetcore/blob/f96dce6889fe67aaed33f0c2b147b8b537358f1e/src/Security/Authentication/Certificate/src/CertificateAuthenticationHandler.cs#L243

OpenIdConnect actually has an issuer concept built in, that's what's used to validate the token and create the Claims.
https://github.com/dotnet/aspnetcore/blob/971df4cbdc045c879f288b8cdfd726fd4212f882/src/Security/Authentication/OpenIdConnect/src/OpenIdConnectHandler.cs#L1239-L1248

Since the validation and claims creation steps are linked together within IdentityModel, that's not something we can change. The closest we could manage is to re-write the identity after it's created.

Why do you need to change the issuer?

**Comment by amoszheng** [comment]
@Tratcher we wanted to use the OpenIdConnect client to implement external login integration for an identity provider that is not yet supported neither by Microsoft out of the box nor [aspnet-contrib](https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers). Since every supported provider changes the issuer of the claims to the name of the external provider (e.g. "Google", "Facebook", "LinkedIn"), we thought we would do the same with the OpenIdConnect client and have our external identity provider integrate seamlessly into the existing pipeline.

Having the OpenIdConnect client support changing the issuer of the claims would allow integration with any identity provider that supports OpenId Connect without having to write a custom provider. That provides a lot of flexibility to integrate new login providers to our applications, now that OpenId connect is pretty much the norm everywhere for this kind of stuff.

**Comment by Tratcher** [comment]
The claims issuer is still set to a unique provider ID like "https://myprovider.com", right? Is your concern about the format?

**Comment by amoszheng** [comment]
@Tratcher yes, we tried to set it to "MyProvider" but could not do it like it is for the other login providers.

**Comment by Tratcher** [comment]
Can you clarify why you _need_ it to be formatted as "MyProvider" instead of the current value "https://myprovider.com/"?

**Comment by brockallen** [comment]
> Can you clarify why you need it to be formatted as "MyProvider"

I'm guessing because it matches the scheme name, not the issuer. I was always confused by the discrepancy between the social providers and how the OIDC one is done.

**Comment by amoszheng** [comment]
@brockallen you are right, for all other social login providers the issuer matches the scheme name, e.g. for Google the issuer of the claims is "Google", for Facebook it is "Facebook", for LinkedIn it is "LinkedIn", and so forth.

**Comment by Tratcher** [comment]
@brockallen mainly because IdentityModel doesn't provide an API to set the claims issuer independently, and this is the first request I'm aware of to do so. We can file an issue over there to add such an API.

**Comment by Tratcher** [comment]
Filed https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/issues/1972.

**Comment by Tratcher** [comment]
Workaround:
```
    private class IssuerFixupAction : ClaimAction
    {
        public IssuerFixupAction() : base(ClaimTypes.NameIdentifier, string.Empty) { }

        public override void Run(JsonElement userData, ClaimsIdentity identity, string issuer)
        {
            var oldClaims = identity.Claims.ToList();
            foreach (var claim in oldClaims)
            {
                identity.RemoveClaim(claim);
                identity.AddClaim(new Claim(claim.Type, claim.Value, claim.ValueType, issuer, claim.OriginalIssuer, claim.Subject));
            }
        }
    }


        services.AddAuthentication(...).AddCookie().AddOpenIdConnect(o =>
        {
           // ...
            o.ClaimsIssuer = "MyCustomIssuer";
            o.ClaimActions.Add(new IssuerFixupAction());
```

**Comment by amoszheng** [comment]
> Workaround:
> 
> ```
>     private class IssuerFixupAction : ClaimAction
>     {
>         public IssuerFixupAction() : base(ClaimTypes.NameIdentifier, string.Empty) { }
> 
>         public override void Run(JsonElement userData, ClaimsIdentity identity, string issuer)
>         {
>             var oldClaims = identity.Claims.ToList();
>             foreach (var claim in oldClaims)
>             {
>                 identity.RemoveClaim(claim);
>                 identity.AddClaim(new Claim(claim.Type, claim.Value, claim.ValueType, issuer, claim.OriginalIssuer, claim.Subject));
>             }
>         }
>     }
> 
> 
>         services.AddAuthentication(...).AddCookie().AddOpenIdConnect(o =>
>         {
>            // ...
>             o.ClaimsIssuer = "MyCustomIssuer";
>             o.ClaimActions.Add(new IssuerFixupAction());
> ```

Thanks for the workaround! Will try it while we wait for the permanent fix. Keep up the awesome work you guys do in the .NET team!
